### ISSUE TITLE
Potential energy inconsistency in CPU platform

### ISSUE THREAD

**Issue by ajsilveira** [openmm/openmm#2521]
I was checking the derivative calculation for a CustomNonbondedForce when I came across an inconsistency in the LJ energy computed in the CPU platform when this extra potential is added to the system. The following script reproduces the issue.
The pdb and force field files can be found here https://gist.github.com/ajsilveira/ac86ce5d407320a0be8bdeb319715f0a
We are interested in using the derivative as a collective variable. Given that the derivative can be retrieved by using getState(), how easy would it be to modify the CustomCVForce to add the derivative as a collective variable? 
```
from simtk.openmm import app
import simtk.openmm as openmm
from simtk import unit
import pandas as pd

pdb = app.PDBFile('benzene.pdb')
forcefield = app.ForceField('benzene.xml')
system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.PME, nonbondedCutoff=0.5*unit.nanometers)

softcore_lj = "4*lambda_vdw*epsilon*(1-x)/x^2; x=(r/sigma)^6+0.5*(1-lambda_vdw); epsilon=sqrt(epsilon1*epsilon2); sigma=0.5*(sigma1+sigma2)"
softcore = openmm.CustomNonbondedForce(softcore_lj)
softcore.addGlobalParameter('lambda_vdw', 1.0)
softcore.addPerParticleParameter('sigma')
softcore.addPerParticleParameter('epsilon')
for force in system.getForces():
    if isinstance(force, openmm.NonbondedForce):
        for index in range(force.getNumParticles()):
            (q, sig, eps) = force.getParticleParameters(index)
            softcore.addParticle([sig, eps])
softcore.addInteractionGroup([0], [12])
softcore.addEnergyParameterDerivative('lambda_vdw')
softcore.setForceGroup(1)

for force in system.getForces():
    if isinstance(force, openmm.NonbondedForce):
        force.setForceGroup(2)
        force.setReciprocalSpaceForceGroup(3)
print('Without softcore')
frames = []
for name in ['Reference', 'CPU', 'OpenCL', 'CUDA']:
    platform = openmm.Platform.getPlatformByName(name)
    integrator = openmm.CustomIntegrator(0)
    context = openmm.Context(system, integrator, platform)
    context.setPositions(pdb.positions)
    df = pd.DataFrame(data={
        'Softcore': [context.getState(getEnergy=True, groups=2**1).getPotentialEnergy()/unit.kilojoules_per_mole],
        'Lennard-Jones': [context.getState(getEnergy=True, groups=2**2).getPotentialEnergy()/unit.kilojoules_per_mole],
        'Coulomb': [context.getState(getEnergy=True, groups=2**3).getPotentialEnergy()/unit.kilojoules_per_mole],
    }, index=[name])
    frames.append(df)
print(pd.concat(frames))
print('')
print('With softcore')
system.addForce(softcore)
frames = []
for name in ['Reference', 'CPU', 'OpenCL', 'CUDA']:
    platform = openmm.Platform.getPlatformByName(name)
    integrator = openmm.CustomIntegrator(0)
    context = openmm.Context(system, integrator, platform)
    context.setPositions(pdb.positions)
    df = pd.DataFrame(data={
        'Derivative': [context.getState(getParameterDerivatives=True, groups=2**1).getEnergyParameterDerivatives()['lambda_vdw']],
        'Softcore': [context.getState(getEnergy=True, groups=2**1).getPotentialEnergy()/unit.kilojoules_per_mole],
        'Lennard-Jones': [context.getState(getEnergy=True, groups=2**2).getPotentialEnergy()/unit.kilojoules_per_mole],
        'Coulomb': [context.getState(getEnergy=True, groups=2**3).getPotentialEnergy()/unit.kilojoules_per_mole],
    }, index=[name])
    frames.append(df)
print(pd.concat(frames))
```

**Comment by peastman** [comment]
Thanks!  I'll take a look.

> Given that the derivative can be retrieved by using getState(), how easy would it be to modify the CustomCVForce to add the derivative as a collective variable?

To apply forces, you need the derivative of each collective variable with respect to the particle coordinates.  So if you want to use the derivative dE/dC as a collective variable, we would need to implement the calculation of the second derivative d(dE/dC)/dx.  Or equivalently you could think of it as dF/dC, the derivative of the vector of forces on particles with respect to the collective variable.  That would be complicated to implement, and might be slow to compute.

**Comment by jchodera** [comment]
@ajsilveira : While we can't easily use derivatives as collective variables, we can probably use finite-difference derivatives as collective variables. For example, in a CustomIntegrator, we can compute `C = [U(lambda+dlambda) - U(lambda-dlambda)]` and compute forces that depend on `dC/dx`.

**Comment by peastman** [comment]
I just submitted a pull request with the fix: #2544.

The error was triggered because you specified PME as the nonbonded method for the NonbondedForce, but left the CustomNonbondedForce using the default NoCutoff.  I assume that wasn't what you intended to do?  You probably want to add the line

```python
softcore.setNonbondedMethod(CustomNonbondedForce.CutoffPeriodic)
```
